Core/tests/Unit/ExponentialBackoffTest.php

```diff
@@ -160,4 +160,58 @@ public function delayProvider()
             [10, 60000000, 60000000]
         ];
     }
+
+    public function testOnRetryExceptionFunction()
+    {
+        $args = ['foo' => 'bar'];
+        $onRetryExceptionFunction = function (
+            $ex,
+            $retryAttempt,
+            $arguments
+        ) {
+            $message = sprintf('%s, %s', $retryAttempt, $arguments[0]['foo']);
+            throw new \Exception($message);
+        };
+
+        $result = '';
+        $expected = '0, bar';
+        $backoff = new ExponentialBackoff(null, null, $onRetryExceptionFunction);
+        try {
+            $backoff->execute(
+                function ($options) use ($args) {
+                    throw new \Exception();
+                },
+                [$args]
+            );
+        } catch (\Exception $ex) {
+            $result = $ex->getMessage();
+        }
+
+        $this->assertEquals($expected, $result);
```

Comment:
Instead of comparing a string, shall we just do 2 asserts inside the onRetryExceptionFunction`?
The 2 asserts being
- The attempt being equal to 0
- The arguments[0] having the required structure.

Reply:
For that we would be required to pass the `$this` reference as follows:
```php
$onRetryExceptionFunction = function (
    $ex,
    $retryAttempt,
    $arguments
) use($this) {
    $message = sprintf('%s, %s', $retryAttempt, $arguments[0]['foo']);
    throw new \Exception($message);
};
```

I didn't want to add extra scope to this callback thus used this method of getting message from the exception. Should we do the assert thing inside callback?

Reply:
Thanks for this info 💯 .

Applied in commit [ba03c51](https://github.com/googleapis/google-cloud-php/pull/5953/commits/ba03c51d90bb70cd400d755623229350d97ccd2d)